[cpplings] Add template programming exercises

XueSongTap opened the pull request:
### What problem were solved in this pull request?

Issue Number:  #523

Problem:
- Missing C++ template programming exercises, affecting contributors' understanding and application of C++ templates
- Need to enhance the cpplings exercise collection to support the goal of Issue #508

### What is changed and how it works?

- Added template_basics.cpp file, implementing:
  - Basic function template examples showing how to create generic functions
  - Class template implementation and usage, demonstrating definition and usage of generic classes
  - Template specialization techniques with a specialized version for string type
  - Recursive implementation of variadic templates, showing how to handle variable parameter lists
  - SFINAE technique application, selecting different function implementations based on type traits

- Added comprehensive comments explaining template programming concepts and use cases
- Included complete test code to ensure examples run correctly

### Other information
- Run the test case with `./bin/cpplings template_basics`
- Parent Issue: #508 (add more cpplings case)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/oceanbase/miniob?pullRequest=526) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/oceanbase/miniob?pullRequest=526) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/oceanbase/miniob?pullRequest=526) it.</sub>

nautaa commented on src/cpplings/template.cpp:
```diff
@@ -0,0 +1,131 @@
+  return first + sum(args...);
+}
+
+// 5. SFINAE技术示例
```
SFINAE 可以多加些描述，对于没接触过的人来说，可能只看这例子也不明所以。

XueSongTap commented:
H i @nautaa Thank you for your review! I have enhanced the SFINAE documentation:
1. Fixed the incomplete comments
2. Added more detailed explanations of the SFINAE concept
3. Added explanatory comments to the test code

Please take another look and let me know if you have any other suggestions.

nautaa commented on src/cpplings/template.cpp:
```diff
@@ -0,0 +1,139 @@
+
+  // 5. 测试SFINAE
+  std::cout << "\nSFINAE examples:" << std::endl;
+  // - 对于整数类型，调用的是以">=0"判断正负的版本
```
">=0" -> ">0"?

nautaa commented on src/cpplings/template.cpp:
```diff
@@ -0,0 +1,139 @@
+  // - 对于整数类型，调用的是以">=0"判断正负的版本
+  std::cout << "is_positive(42): " << (is_positive(42) ? "true" : "false") << std::endl;
+  std::cout << "is_positive(-42): " << (is_positive(-42) ? "true" : "false") << std::endl;
+  // - 对于浮点类型，调用的是用">0"判断正负的版本
```
">0" -> ">0.0" ?

XueSongTap commented:
Hi @nautaa,

Thank you for your review! I have enhanced the SFINAE documentation:

- Fixed the comparison operator comments (">= 0" to "> 0", "> 0" to "> 0.0")

Please take another look and let me know if you have any other suggestions.
